src/io/dataloader.cc

```diff
@@ -170,8 +164,7 @@ class ThreadedDataLoader : public IIterator<TBlobBatch> {
   /*! \brief batched buffer */
   std::vector<NDArray> batched_buffer_;
   /*! \brief pointer to dataset */
-  // std::shared_ptr<Dataset> dataset_;
-  std::vector<std::shared_ptr<Dataset>> datasets_;
+  std::shared_ptr<Dataset> dataset_;
```

Comment:
Would we have some double free now? Because this shared_ptr would clean up the Dataset at the end of it's lifetime, and we also have a MXDatasetFree C API

Reply:
Nevermind, the MXDatasetCreate actually creates a shared pointer on the heap. So we're just reusing that and there is no double free here. `*out = new std::shared_ptr<Dataset>(dataset); `